Adds blog on patching a Groq client instance to add tracing support + some general principles on tracing customization

djliden opened the pull request.

BenWilson2 commented on website/blog/2024-12-16-custom-tracing-provider/index.md:
```diff
@@ -0,0 +1,376 @@
+
+A few points to note:
+
+- We are wrapping a method on a client *instance*, not a class. This is a fairly lightweight approach that does what we need it to do without requiring changes to the Groq SDK code.
```
Might want to use the term "patching" instead of wrapping. We are still guerilla patching here by overriding the existing implementation. Wrapping would be more of using the instance of the return obj + method reference of the return value of trace_groq_chat.

BenWilson2 commented on website/blog/2024-12-16-custom-tracing-provider/index.md:
```diff
@@ -0,0 +1,376 @@
+
+### Step 3: Wrap the instance method and try it out
+
+Now that we have a tracing decorator, we can wrap the `chat.completions.create` method on a Groq client instance and try it out.
```
"apply our patch" instead of "wrap"

BenWilson2 commented on website/blog/2024-12-16-custom-tracing-provider/index.md:
```diff
@@ -0,0 +1,376 @@
+
+![Tool Calls](./5_tool_call.png)
+
+## Orchestration: Building a tool calling
```
fragment?

BenWilson2 reviewed (commented):
Overall, this is fantastic. Minor nits and comments.

BenWilson2 commented:
cc @B-Step62 @daniellok-db for inputs on tracing messaging for this blog :D

djliden commented on website/blog/2024-12-16-custom-tracing-provider/index.md:
```diff
@@ -0,0 +1,376 @@
+
+[MLflow Tracing](https://mlflow.org/docs/latest/llms/tracing/index.html) is an observability tool in MLflow that captures detailed execution traces for GenAI applications and workflows. In addition to inputs, outputs, and metadata for individual calls, MLflow tracing can also capture intermediate steps such as tool calls, reasoning steps, retrieval steps, or other custom steps.
+
+MLflow provides [built-in Tracing support](https://mlflow.org/docs/latest/llms/tracing/index.html#automatic-tracing) for many popular LLM providers and orchestration frameworks. If you are using one of these providers, you can enable tracing with a single line of code: `mlflow.<provider>.autolog()`. While MLflow's autologging capabilities cover many of the most widely-used LLM providers and orche ...
```
Can you expand on this point? Not sure what this means in practice.

B-Step62 commented:
@BenWilson2 

1. Actually Groq tracing is coming soon most likely in 2.20 ([PR](https://github.com/mlflow/mlflow/pull/14006)). Shall we change the client to sth else, for example, [Ollama Python SDK](https://github.com/ollama/ollama-python)?
2. Also the way we render chat messages on span UI changes significantly in 2.20 by this [PR](https://github.com/mlflow/mlflow/pull/14087). After that, we should NOT recommend modifying the raw input/output when logging it to span, instead use custom attribute `mlflow.chatMessage`. I think this blog is a perfect place to introduce the new method🙂

Incorporating these changes means that we need to wait this until 2.20 release (mid Jan), but I feel we don't want to publish a blog that we know becomes stale in a month.

djliden commented:
Thanks @B-Step62! I'll hold off until January and then revise with the updated approach & a different provider.

djliden commented:
Updated to use the Ollama python SDK as an example and to use the `set_span_chat_messages` and `set_span_chat_tools` utils for formatting the chat pane. Updated all screenshots. Overall narrative is largely the same.

BenWilson2 reviewed (approved):
Looks great!

BenWilson2 commented on website/src/posts.ts:
```diff
@@ -26,6 +26,21 @@ export type Release = {
+      },
+    ],
+    date: "2025-01-30",
+    thumbnail: "img/blog/tracing-new-provider.png",
```
```suggestion
    thumbnail: "/img/blog/tracing-new-provider.png",
```

BenWilson2 commented on website/blog/2025-01-30-custom-tracing-provider/index.md:
```diff
@@ -0,0 +1,516 @@
+tags: [genai, tracing, ollama]
+slug: custom-tracing
+authors: [daniel-liden]
+thumbnail: img/blog/tracing-new-provider.png
```
```suggestion
thumbnail: /img/blog/tracing-new-provider.png
```
